### DISCUSSION TITLE
ERROR: toktx is not present in PATH or TOKTX_PATH is not set

### DISCUSSION THREAD

**Discussion by Bketting** [zeux/meshoptimizer#247] [Q&A] [answered]
Apologies for the novice-level question, but I have little coding experience. I'm trying to convert .jpg's in a .glb file into .ktx2. Installed both gltfpack and KTX-Software v4.0.0-beta6 (the latter is located in Program Files).

When I try to run the command line code "gltfpack -i x.glb -tc -o y.glb" I get the following error:

_ERROR: toktx is not present in PATH or TOKTX_PATH is not set_

I'm running Windows. Running gltfpack without -tc is working fine. What is going wrong?

**Comment by zeux** [comment]
I don't think KTX-Software performs the correct setup of PATH environment variable; you'd have to add `C:\Program Files\KTX-Software\bin` to PATH environment variable or alternatively set TOKTX_PATH environment variable to `C:\Program Files\KTX-Software\bin\toktx.exe`

**Reply by Bketting** [reply]
Tried adding both paths manually as well without success, toktx does work by itself though

**Reply by zeux** [reply]
Assuming you use npm build of gltfpack, to fix the path to gltfpack itself you should make sure `C:\Users\YOURUSERNAME\AppData\Roaming\npm` is in the PATH

**Reply by Bketting** [reply]
Got gltfpack running again, thank you! No luck on the toktx though

**Reply by zeux** [reply]
What does running `set PATH` in a terminal output for you?

**Comment by kartik-madhak** [comment]
I also have the same problem. I have tried adding toktx to PATH and also tried setting TOKTX_PATH to toktx executable. Still it is saying ```Error: toktx is not present in PATH or TOKTX_PATH is not set``` when I write the ```-tc``` flag.

**Reply by zeux** [reply]
What's your operating system? Are you using gltfpack from npm or a native build? I'd recommend trying a native build first (you can download a binary on Releases page) if you're using an npm build. If that works and npm build doesn't, I'd need the exact version of NPM, OS version and the contents of PATH variable to investigate further.

**Reply by kartik-madhak** [reply]
I am running Windows 10 (Version 1909). I tried both through npm and from release page of gltfpack. Both are giving the same error.
Also, my npm version is 6.13.4
and my PATH has this directory listed - C:\Users\<...>\3D Objects\KTX-Software\bin;

**Reply by zeux** [reply]
If you just write `toktx` in the terminal and press Enter, does toktx actually run? This doesn't seem like a corect path for KTX-Software unless you installed it there intentionally...

**Reply by zeux** [reply]
Specifically try running `toktx --version`

**Reply by kartik-madhak** [reply]
Yes, I put it there intentionally. Should I try installing it in its usual path?
Also yes ```toktx``` command runs in terminal.

**Reply by zeux** [reply]
Maybe the issue is that TOKTX_PATH overrides it; what does `set TOKTX_PATH` print?

**Reply by kartik-madhak** [reply]
```toktx v4.0.0-beta7~1```
Should I try installing from the stable release?

**Reply by zeux** [reply]
Nah, this one should work. Btw you can also pass `-v` to gltfpack and it will print the command it's executing along with the status code.

**Reply by kartik-madhak** [reply]
```set TOKTX_PATH``` returns ```TOKTX_PATH=C:\Users\<...>\3D Objects\KTX-Software\bin\toktx.exe```

**Reply by zeux** [reply]
Try changing TOKTX_PATH to the same value but quoted with double quotes? Here's what I have atm:

```
TOKTX_PATH="C:\Program Files\KTX-Software\bin\toktx.exe"
```

**Reply by kartik-madhak** [reply]
> Nah, this one should work. Btw you can also pass `-v` to gltfpack and it will print the command it's executing along with the status code.

Do you mean something like this?
```gltfpack -v -tc -i DamagedHelmet.gltf -o trial.glb```

**Reply by zeux** [reply]
> Do you mean something like this?

Sorry, I meant `-vv`. So yeah `gltfpack -vv -tc -i DamagedHelmet.gltf -o trial.glb`

**Reply by kartik-madhak** [reply]
> > Do you mean something like this?
> 
> Sorry, I meant `-vv`. So yeah `gltfpack -vv -tc -i DamagedHelmet.gltf -o trial.glb`

It returned 
```
C:\Users\<...>\3D Objects\KTX-Software\bin\toktx.exe --version => 1
basisu -version => 1
Error: toktx is not present in PATH or TOKTX_PATH is not set
```

**Reply by zeux** [reply]
Yeah so I suspect this is a quoting problem then. See https://github.com/zeux/meshoptimizer/discussions/247#discussioncomment-506661

**Reply by kartik-madhak** [reply]
> Yeah so I suspect this is a quoting problem then. See [#247 (reply in thread)](https://github.com/zeux/meshoptimizer/discussions/247#discussioncomment-506661)

I will try setting with the quote. But I dont know why there is a problem because I set it initially using the "Edit System Environment Variables" GUI.

**Reply by zeux** [reply]
It's just that right now gltfpack doesn't auto-quote the values with spaces. If this is indeed the problem I'll look into fixing this.
![image](https://user-images.githubusercontent.com/1106629/111876493-3985d200-895c-11eb-9d39-5a06e3958c72.png)

**Reply by kartik-madhak** [reply]
Thanks, It worked with the explicit double quotes. Also, I had to remove toktx bin directory from PATH.

**Reply by zeux** [reply]
Alright - thanks! I'll look into improving the code here as well as the diagnostics.

**Comment by zeux** [comment, marked as answer]
I've opened a PR that should solve the issues with various setups reported in this topic, thanks a lot for the feedback! https://github.com/zeux/meshoptimizer/pull/264

**Comment by Habacef** [comment]
sam problem here, using it inside of a docker, which makes it far more complicated

**Comment by kitaedesigns** [comment]
This problem seems to still be unresolved, I just tried -tc and -te with gltfpack and it returns the same error.
Error: toktx is not present in PATH or TOKTX_PATH is not set
